### ISSUE TITLE
Change the execution method of main

### ISSUE THREAD

**Issue by AntonisZks** [AntonisZks/Information-Systems-Software-Development-Project#57]
I suggest we make our **main** file execution more flexible by changing some, and adding extra flags to the arguments. Specifically what I suggest is to use key-value pairs for the arguments. For example if the user wants to enter the `k` parameter for the search process to be `100`, then they have to call the program like that:
```bash
./bin/main <some-arguments> -k 100 <some-other-arguments>
```
By doing the above we give the user a more flexible way of interacting with the application, as they don't have to remember which argument goes first every time, but instead just typing the key of the argument they want to enter, and then enter its value. 

Before analyzing the key-flags I have came across, I will describe how I think the application should be transformed. The **main** function will force the user to enter as a first argument the mode that they want to run the application. Specifically there will be two modes of the application, **Create** and **Test**.

1. **Create:** This mode will have a flag `--create` and it goinf to tell the application to create a **Vamana Index** according to the next **user-defined flags**. These flags will be the **parameters** for creating the **Vamana Index** and specifically:
    
    - The key `-base-file` refers to the **path** of the **base vectors file**.
    - The key `-L` refers to the **L** parameter of the **search process**.
    - The key `-R` refers to the **R** parameter of the **search process**.
    - The key `-alpha` refers to the **$\alpha$** parameter of the **search process**.
    - The key `-save` refers to the **filename** we want to give for the **vaman index** graph, if we want to save it.
  
> [!IMPORTANT] 
> The `-save` flag-key will not necessary to enter. This gives the user the option to save the vamana index only if they want to.

> [!TIP]
> An example of using the **create** mode of the application can be something like the following:
> ```bash
> ./bin/main --create -L 120 -R 14 -alpha 1.0 -save 'index.vig'
> ```

2. **Test:** This mode will have a flag `--test`, and it's going to test a specific **Vamana Index** and evaluate it. Instead of speding too much time on creating a Vamana Index and then test it, this mode will give the user the ability to **load** an **existing** vamana index just by entering the **filename** that it's been saved in, and then the application will **load** the contents of the corresponding **binary file** and create the graph. After that, given a **number of arguments** that specify how the **search** is going to be executed, the application performs a **searching process** in the graph, **finds** the **approximate nearest neighbors** of a given **query vector** and **evaluates** the result using the given groundtruth file. The flag-keys I suggest building are the following:

    - The key `-load` refers to the **path** of the **binary file** containing the **Vamana Index Graph** that we want to test.
    - The key `-k` refers to the **k** parameter of the **search process**
    - The key `-L` refers to the **L** parameter of the **search process**.
    - The key `-gt-file` refers to the **path** of the **groundtruth file** we want to enter.
    - The key `-query-file` refers to the **path** of the **query vectors file**.
    - The key `-query` refers to the **index** of the **query vector** we want to test.

> [!TIP]
> An example of using the **test** mode of the application can be the following:
> ```bash
> ./bin/main --test -load 'index.vig' -k 100 -L 120 -gt-file 'siftsmall_groundtruth.ivecs' -query-file 'siftsmall_query.fvecs' -query 0
> ```

In summary a complete usage for the application, according to the above would be something like the following:
```bash
> user@linux-desktop:-$ ./bin/main --create -L 120 -R 14 -alpha 1.0 -save 'index.vig'
Vamana Index created after 12m 13s.
Vamana Index was saved successfully to 'index.vig'
> user@linux-desktop:-$
> user@linux-desktop:-$ ./bin/main --test -load 'index.vig' -k 100 -L 120 -gt-file 'siftsmall_groundtruth.ivecs' -query-file 'siftsmall_query.fvecs' -query 0
Search process took 5s to run
Recall Evaluation: 99%
```
These are my ideas for future work on this application. I can work on the above. Let me know what you think.

**Comment by StavrosJKw** [comment]
Incredible work @AntonisZks ! It responds to some of the **issues** we talked about and faced in the execution of the first part of the project. 

>[!Tip]
>For **demo** purposes, we could create a directory `/demos`, in which we can add a variety of the vamana index graphs you suggested above, that are ready to use for demonstration purposes.
>
>An example of the demo files could be: 
>- `demo_small.vig`
>- `demo_big.vig`
>- `demo_complex.vig`
>- etc..
>
>And they will be used in the way you suggested:
>```
>./bin/main --test -load 'demo.vig' -k 100 -L 120 -gt-file 'siftsmall_groundtruth.ivecs' -query-file 'siftsmall_query.fvecs' -query 0
>```

>[!important]
>It is of course implied that we will extensively documented each demos contents, purpose and what it tests/show in the projects README.md file.

We could also create a dedicated mode with the corresponding flag `--demo`,  but I believe this will give us less flexibility and less variety in the execution.

**Comment by AntonisZks** [comment]
I have done an implementation of the above idea inside my branch.
